fabric-chat-api-v1/src/main/java/net/fabricmc/fabric/api/chat/v1/ChatDecoratorEvent.java

```diff
@@ -0,0 +1,101 @@
+ * function. If not given, the chat decorator will run in the default phase, which is between
+ * the content phase and the styling phase.
+ *
+ * <p>When implementing a chat decorator, it is <strong>very important that the decorator returns
+ * the same contents between previewing and submission</strong> - otherwise the game discards the
+ * message because it was improperly signed. The most straightforward way of doing this is by
+ * making sure the decorator returns the same result for a given message. Another way to solve this
+ * issue is to cache the decorated message.
+ *
+ * <p>Example of registering a content phase chat decorator:
+ *
+ * <pre><code>
+ * ChatDecoratorEvent.EVENT.register(ChatDecoratorEvent.CONTENT_PHASE, (sender, message) -> {
+ *     // Add smiley face. Has to shallowCopy() to get a MutableText with siblings and styles.
+ *     return message.shallowCopy().append(" :)");
+ * });
+ * </code></pre>
+ *
+ * <p>Example of registering a styling phase chat decorator:
+ *
+ * <pre><code>
+ * ChatDecoratorEvent.EVENT.register(ChatDecoratorEvent.STYLING_PHASE, (sender, message) -> {
+ *     // Apply orange color to messages sent by server operators
+ *     if (sender != null && sender.server.getPlayerManager().isOperator(sender.getGameProfile())) {
+ *         return CompletableFuture.completedFuture(
+ *             message.shallowCopy().styled(style -> style.withColor(0xFFA500)));
+ *     }
+ *     return CompletableFuture.completedFuture(message);
+ * });
+ * </code></pre>
+ */
+public final class ChatDecoratorEvent {
+	private ChatDecoratorEvent() {
+	}
+
+	/**
+	 * The content phase of the event, passed when registering a chat decorator. Use this when
+	 * the chat decorator modifies the text content of the message.
+	 */
+	public static final Identifier CONTENT_PHASE = new Identifier("fabric-chat-api-v1", "content");
+	/**
+	 * The styling phase of the event, passed when registering a chat decorator. Use this when
+	 * the chat decorator only modifies the styling of the message with the text intact.
+	 */
+	public static final Identifier STYLING_PHASE = new Identifier("fabric-chat-api-v1", "styling");
+
+	public static final Event<ChatDecorator> EVENT = EventFactory.createWithPhases(ChatDecorator.class, decorators -> (sender, message) -> {
+		CompletableFuture<Text> future = null;
+
+		for (ChatDecorator decorator : decorators) {
+			if (future == null) {
+				future = decorator.decorate(sender, message).thenApply(ChatDecoratorEvent::validateDecoratorResult);
+			} else {
+				future = future.thenCompose((decorated) -> decorator.decorate(sender, decorated).thenApply(ChatDecoratorEvent::validateDecoratorResult));
+			}
+		}
+
+		return future == null ? CompletableFuture.completedFuture(message) : future;
+	}, CONTENT_PHASE, Event.DEFAULT_PHASE, STYLING_PHASE);
+
+	private static <T extends Text> T validateDecoratorResult(T value) {
+		return Objects.requireNonNull(value, "chat decorator must not return null");
```

Comment:
Is it worth including the ChatDecorator classname in this? I can see it beign a slight pain to figure out what is causing it?